Local function in struct accessing this (error CS1673)

**Describe the bug**

The standard should forbid a local function in a member of a `struct` type from accessing `this` or instance members of `this`.  I don't see where the C# 7.x draft does that.

**Example**

The following should not be allowed.  [SharpLab](https://sharplab.io/#v2:CYLg1APgzgLgTgVwMYwAQGVUG8CwAoVQ1ASwDs1iBufI1GogAQBZUBZACgEpt7bDMAclx4E+YhgHZUMABbEolQgHolqAKZw4AezioAwugCMANgkBmXnwC+l2raJk0AeWG5RY2pJKKV6zTv0jUwt3a0sbPAigA===)

```csharp
struct S {
    int i;
    
    void M() {
        S N() {
            return this;  // error CS1673
        }
        
        int O() {
            return i; // error CS1673
        }
    }
}
```

**Expected behavior**

In [§13.6.4 (Local function declarations)](https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/statements.md#1364-local-function-declarations), declare that a local function in a member of a `struct` shall not access `this` or instance members of `this`.  Static members are okay though.

Add the constraint in or near this paragraph, which talks about variables captured by local functions: <https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/statements.md#L518>

**Additional context**

This is a new problem in C# 7, as the equivalent thing for anonymous delegates and lambda expressions is disallowed in [§12.19.3 (Anonymous function bodies)](https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/expressions.md#12193-anonymous-function-bodies): <https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/expressions.md?plain=1#L4978>

A surprising interaction with `nameof`:

```csharp
struct S {
    int i;
    
    void M() {
        string N() {
            return nameof(this);  // error CS1673
        }

        string O() {
            return nameof(this.i);  // error CS1673
        }

        string P() {
            return nameof(i);  // OK, returns "i"
        }
    }
}
```

*Tracked as <https://github.com/dotnet/csharpstandard/issues/931>*

```csharp

using System;
class DAttribute : Attribute {
    public DAttribute(string s) {}
}
struct S {
    int j;
    void LeA() {
        // OK
        Action<int> a = ([D(nameof(this.j))] int i) => {};
    }
    void LeD() {
        // error CS1065, default value not allowed
        Action<string> action = (string s = nameof(this.j)) => {};
    }
    void LfA() {
        // OK
        void F([D(nameof(this.j))] int i) {}
    }
    void LfD() {
        // error CS1673, cannot access this in local function
        void F(string s = nameof(this.j)) {}
    }
}
```

That seems pretty inconsistent. Why is `nameof(this.j)` allowed in an attribute argument in the parameter list, but not in a default parameter value?  Is that what the standard should specify?

*Tracked as <https://github.com/dotnet/csharpstandard/issues/931>*

Expression trees seem to allow neither attributes nor default values, so the behaviour is more consistent for them:

```csharp
using System;
using System.Linq.Expressions;
class DAttribute : Attribute {
    public DAttribute(string s) {}
}
struct S {
    int j;
    void ExA() {
        // error CS8972, attribute not allowed in expression tree
        Expression<Func<string, string>> e = ([D(nameof(this.j))] string s) => "";
    }
    void ExD() {
        // error CS1065, default value not allowed in expression tree
        Expression<Func<string, string>> e = (string s = nameof(this.j)) => "";
    }
    void ExR() {
        // error CS1673, lambda expression in struct cannot access this
        Expression<Func<string, string>> e = (string s) => nameof(this.j);
    }
}
```

--- a/standard/statements.md
+++ b/standard/statements.md
@@ -519,7 +519,9 @@ A local function is declared at block scope, and that function may capture varia
 
 When the type of `this` is a struct type, it is a compile-time error for the body of a local function to access `this`. This is true whether the access is explicit (as in `this.x`) or implicit (as in `x` where `x` is an instance member of the struct). This rule only prohibits such access and does not affect whether member lookup results in a member of the struct.
 
-> *Note*: this rule mirrors the rule for anonymous functions in [§12.19.3](expressions.md#12193-anonymous-function-bodies). *end note*
+It is a compile-time error for the body of the local function to contain a `goto` statement, a `break` statement, or a `continue` statement whose target is outside the body of the local function.
+
+> *Note*: the above rules for `this` and `goto` mirror the rules for anonymous functions in [§12.19.3](expressions.md#12193-anonymous-function-bodies). *end note*
 
 A local function may be called from a lexical point prior to its declaration. However, it is a compile-time error for the function to be declared lexically prior to the declaration of a variable used in the local function ([§7.7](basic-concepts.md#77-scopes)).
 

--- a/standard/statements.md
+++ b/standard/statements.md
@@ -517,6 +517,10 @@ A *local_function_declaration* may include one `async` ([§15.15](classes.md#151
 
 A local function is declared at block scope, and that function may capture variables from the enclosing scopes. It is a compile-time error if a captured variable is read by the body of the local function but is not definitely assigned before each call to the function. The compiler shall determine which variables are definitely assigned on return ([§9.4.4.33](variables.md#94433-rules-for-variables-in-local-functions)).
 
+When the type of `this` is a struct type, it is a compile-time error for the body of a local function to access `this`. This is true whether the access is explicit (as in `this.x`) or implicit (as in `x` where `x` is an instance member of the struct). This rule only prohibits such access and does not affect whether member lookup results in a member of the struct.
+
+> *Note*: this rule mirrors the rule for anonymous functions in [§12.19.3](expressions.md#12193-anonymous-function-bodies). *end note*
+
 A local function may be called from a lexical point prior to its declaration. However, it is a compile-time error for the function to be declared lexically prior to the declaration of a variable used in the local function ([§7.7](basic-concepts.md#77-scopes)).
 
 It is a compile-time error for a local function to declare a parameter, type parameter or local variable with the same name as one declared in any enclosing local variable declaration space.

